File: src/leetCode/c++/Solution1851.cpp
//Intervals Heap O(nlogn + qlogq) O(n + q), n = intervals.size(), q = queries.size()
class Solution1851 {
public:
    vector<int> minInterval(vector<vector<int>>& intervals, vector<int>& queries) {
        vector<pair<int, int>> ques;//<<數字, 原數組的位置>, ...>
        for(int i=0; i<queries.size(); ++i)
            ques.push_back({queries[i], i});

        sort(intervals.begin(), intervals.end());//依照start小到大
        sort(ques.begin(), ques.end());//依照數字小到大
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;//<長度, end>, 依長度小到大
        int i = 0;
        vector<int> res(queries.size());
        for(auto que : ques){
            while(i < intervals.size() && intervals[i][0] <= que.first){//加入範圍內
                pq.push({intervals[i][1] - intervals[i][0] + 1, intervals[i][1]});
                ++i;
            }

            while(!pq.empty() && pq.top().second < que.first){
                pq.pop();//移除範圍外
            }

            res[que.second] = pq.empty() ? -1 : pq.top().first;
        }
        return res;
    }
};
/* intervals = [[2,3],[2,5],[1,8],[20,25]], queries = [2,19,5,22]

            s   e   len
q=2 heap=   2   3   2
            2   5   4
            1   8   8
q=5 heap=   2   5   4
            1   8   8
q=19heap=
q=22heap=   20  25  5
*/

File: java/Solution1851.java
package leetCode.java;

import java.util.*;

//Intervals Heap O(nlogn + qlogq) O(n + q), n = intervals.length, q = queries.length
class Solution1851 {
    public int[] minInterval(int[][] intervals, int[] queries) {
        // Offline Query, 先記錄queries index, 之後計算完再重新放回原位置
        int n = queries.length;
        int[][] qs = new int[n][2]; // [數字][原queries index] 
        for (int i = 0; i < n; ++i) {
            qs[i][0] = queries[i];
            qs[i][1] = i;
        }

        // 排序依照start與qs, 方便配對
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]); // 依照start小到大
        Arrays.sort(qs, (a, b) -> a[0] - b[0]); // 依照queries數字小到大
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]); // <[長度][right index]>, 依照長度小到大

        int i = 0;
        int[] res = new int[n];
        for (int[] q : qs) {
            int query = q[0];
            int originalIndex = q[1];

            // 尋找start<=q的值, 放入篩選
            while (i < intervals.length && intervals[i][0] <= query) {
                pq.offer(new int[]{intervals[i][1] - intervals[i][0] + 1, intervals[i][1]}); // [長度][right index]
                ++i;
            }

            // 把right<q的值移出
            while (!pq.isEmpty() && pq.peek()[1] < query) {
                pq.poll();
            }

            res[originalIndex] = pq.isEmpty() ? -1 : pq.peek()[0]; // peek()[0]=最短值
        }
        return res;
    }
}
/* intervals = [[2,3],[2,5],[1,8],[20,25]], queries = [2,19,5,22]


        heap(len小到大):

        s   e   len
q[0]=2  2   3   2
        2   5   4
        1   8   8

---------------------
q[0]=5  2   5   4
        1   8   8

---------------------
q[0]=19

---------------------
q[0]=22 20  25  5


*/